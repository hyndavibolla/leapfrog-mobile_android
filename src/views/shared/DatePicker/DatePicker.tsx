import React, { ComponentProps, memo, useCallback, useState } from 'react';
import DateTimePickerModal from 'react-native-modal-datetime-picker';
import moment from 'moment';

import { Input } from '../Input';
import { useTestingHelper } from '../../../utils/useTestingHelper';

export interface Props
  extends Pick<
    ComponentProps<typeof DateTimePickerModal>,
    Exclude<keyof ComponentProps<typeof DateTimePickerModal>, 'onConfirm' | 'onCancel' | 'isVisible' | 'mode'>
  > {
  isInvalid?: boolean;
  onChangeDate: (date: Date) => void;
}

export default memo(({ isInvalid, date, onChangeDate, ...props }: Props) => {
  const { getTestIdProps } = useTestingHelper('date-picker');
  const [isDatePickerVisible, setDatePickerVisibility] = useState(false);

  const showDatePicker = useCallback(() => setDatePickerVisibility(true), []);
  const hideDatePicker = useCallback(() => setDatePickerVisibility(false), []);
  const onConfirm = useCallback(
    (d: Date) => {
      onChangeDate(d);
      hideDatePicker();
    },
    [onChangeDate, hideDatePicker]
  );
  const formattedDate = date ? moment(date).format('MMM Do, YYYY') : '';

  return (
    <>
      <Input value={formattedDate} onPressIn={showDatePicker} isInvalid={isInvalid} showSoftInputOnFocus={false} {...getTestIdProps('input')} />
      <DateTimePickerModal {...props} isVisible={isDatePickerVisible} mode="date" onConfirm={onConfirm} onCancel={hideDatePicker} />
    </>
  );
});
