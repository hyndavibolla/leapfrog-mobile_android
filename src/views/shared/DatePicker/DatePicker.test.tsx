import React from 'react';

import DatePicker, { Props } from './DatePicker';
import { fireEvent, render } from '@testing-library/react-native';

describe('DatePicker', () => {
  let props: Props;
  beforeEach(() => {
    props = {
      date: new Date(1980),
      onChangeDate: jest.fn()
    };
  });

  it('it should render', () => {
    const { toJSON } = render(<DatePicker {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('it should render invalid', () => {
    const { toJSON } = render(<DatePicker {...props} date={null} isInvalid={true} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('it should pick a date', () => {
    const date = new Date(1980);
    const { getByTestId } = render(<DatePicker {...props} date={null} isInvalid={true} testID="picker" />);
    fireEvent(getByTestId('date-picker-input'), 'onPressIn');
    fireEvent(getByTestId('picker'), 'onConfirm', date);
    expect(props.onChangeDate).toBeCalledWith(date);
  });
});
