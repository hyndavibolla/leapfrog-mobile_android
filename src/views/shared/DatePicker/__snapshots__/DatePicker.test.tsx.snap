// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker it should render 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "elevation": 3,
          "shadowColor": "#262C2F",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 4,
          "borderBottomRightRadius": 4,
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
          "width": "100%",
        },
        false,
      ]
    }
  >
    <TextInput
      accessibilityLabel="date-picker-input"
      accessible={true}
      onChangeText={[Function]}
      onPressIn={[Function]}
      showSoftInputOnFocus={false}
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderBottomLeftRadius": 4,
            "borderBottomRightRadius": 4,
            "borderColor": "transparent",
            "borderStyle": "solid",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
            "borderWidth": 1,
            "paddingBottom": 15,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
      testID="date-picker-input"
      value="Dec 31st, 1969"
    />
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  />,
]
`;

exports[`DatePicker it should render invalid 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "elevation": 3,
          "shadowColor": "#262C2F",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 4,
          "borderBottomRightRadius": 4,
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
          "width": "100%",
        },
        false,
      ]
    }
  >
    <TextInput
      accessibilityLabel="date-picker-input"
      accessible={true}
      isInvalid={true}
      onChangeText={[Function]}
      onPressIn={[Function]}
      showSoftInputOnFocus={false}
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderBottomLeftRadius": 4,
            "borderBottomRightRadius": 4,
            "borderColor": "#E02020",
            "borderStyle": "solid",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
            "borderWidth": 1,
            "paddingBottom": 15,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 15,
            "width": "100%",
          },
          undefined,
        ]
      }
      testID="date-picker-input"
      value=""
    />
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  />,
]
`;
